common/clientFactory.go

```diff
@@ -0,0 +1,106 @@
+// Copyright © 2017 Microsoft <wastore@microsoft.com>
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package common
+
+import (
+	"errors"
+	"fmt"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blob"
+	blobservice "github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/service"
+)
+
+// newClientCallbacks is a Generic Type to allow client creation error handling to live in a single place (createClient)
+// T = Client type
+// U = SharedKeyCredential type
+// Note : Could also make azcore.ClientOptions generic here if one day different storage service clients have additional options. This would also make the callback definitions easier.
+type newClientCallbacks[T, U any] struct {
+	TokenCredential        func(string, azcore.TokenCredential, azcore.ClientOptions) (*T, error)
+	NewSharedKeyCredential func(string, string) (*U, error)
+	SharedKeyCredential    func(string, *U, azcore.ClientOptions) (*T, error)
+	NoCredential           func(string, azcore.ClientOptions) (*T, error)
+}
+
+// createClient is a generic method to allow client creation error handling to live in a single place
+func createClient[T, U any](callbacks newClientCallbacks[T, U], u string, credInfo CredentialInfo, credOpOptions CredentialOpOptions, options azcore.ClientOptions) (*T, error) {
+	switch credInfo.CredentialType {
+	case ECredentialType.OAuthToken():
+		if credInfo.OAuthTokenInfo.IsEmpty() {
+			credOpOptions.panicError(errors.New("invalid state, cannot get valid OAuth token information"))
+		}
+		tc, err := credInfo.OAuthTokenInfo.GetTokenCredential()
+		if err != nil {
+			credOpOptions.panicError(fmt.Errorf("unable to get token credential due to reason (%s)", err.Error()))
+		}
+		return callbacks.TokenCredential(u, tc, options)
+	case ECredentialType.SharedKey():
```

Comment:
Shared key should not be required. We do not allow this option.

Reply:
Why is this in current code to create pipelines? Was it just missed to remove at some point?

Are we sure there are no tests or anything that uses SharedKey?

Reply:
Removed SharedKey part